### PR TITLE
Leon and Oleksandr add interactive map

### PR DESCRIPTION
**Pull request by xaanders**
# Description
image

## Related PRS 
This frontend PR is related to the #[599 ](https://github.com/OneCommunityGlobal/HGNRest/pull/599) backend PR.

## Main changes explained:
1. Added interactive map
2. Added new route
3. adding new users(Prior HGN data) functionality
4. editing/removing manually added users functionality
5. changed format of storing and saving location data of HGN users
6. editing HGN users info through map

## How to test:
1. set up backend
2. check out into current branch
3. do `npm install` and `npm run start:local` to run this PR locally
4. log in as owner user
5. go to reports -> team locations
6. add, edit, remove users with marker, list, and search
7. 
a) create new HGN user: and punch at least 10 tangible hours to the user     
b)  use existing one: go to user profile and update its location by getting timezone and saving 
8. find user on the map and try edit through marker/list/search 
9. go to user profile and compare
10. edit some info in the profile and compare to the map
11. login to any other user and confirm there is no way to add/edit/remove/search users

## Screenshots or videos of changes:

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/108314277/6e90cd5f-1b10-4b89-ac45-880db331f864

### CODE REVIEW

**Comment by ptrpengdev** [comment]
Hi @xaanders, it appears that there are code conflicts with the latest code in the development branch. Could you please take a moment to resolve these conflicts? I will review the PR and approve it once the conflict solved. Thank you!

**Comment by xaanders** [comment]
HI @ptrpengdev! Thank you. Done.

**Comment by ptrpengdev** [comment]
Hi @xaanders, I've just tested the PR for the iterative map locally. I encountered some issues, but I wasn't sure if these are specific to my end or not.

**Issues:**
1. Total Countries count is not correct.
2. Search bar and buttons are missing.
3. Not able to edit

**Screenshot:**
image


**Steps to reproduce:**
1. Checkout Frontend branch [leon_add_interactive_map](https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/pull/1503) and backend branch [leon_add_interactive_map](https://github.com/OneCommunityGlobal/HGNRest/pull/599)
2. Run `npm install`
4. Acquire an API key from opencagedata.com
5. Log in with the Admin role account -> Go to Reports -> Team Locations 

**Video:**
https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/69882989/855abdb3-f829-4351-8402-f0d8884f8486

**Comment by xaanders** [comment]
1. Total countries is fixed
2. Only owner is able to do anything ( step 4 )

**Comment by ptrpengdev** [comment]
> 1. Total countries is fixed
> 2. Only owner is able to do anything ( step 4 )

Got it. Will test with the Owner role. Thank you!

**Review by lichanghaogit — COMMENTED** [review]
hi @xaanders 
the map renders perfect on my local environment and the number of countries show correctly this time. when in admin and volunteer login, the edit of team number are prohibited.
Screenshot 2023-11-17 at 3 29 56 PM
Screenshot 2023-11-17 at 3 31 41 PM
when I try to create team on the local machine, it seems the new team is never saved. could you help me check how to create a team to test functions for onwers?
Thank you!

**Comment by ptrpengdev** [comment]
Hi @xaanders, I have tested the functionalities with the Owner role, and everything worked as expected. Great work!

Additionally, I wanted to share my personal thought: it would be great to have tooltips or caption text over the location field of the "Adding New User" form in the map's "Add Person" feature as an enhancement. I encountered difficulty using the "get location" function. I tried entering locations manually like Boston, MA, California, USA, and copying locations from existing users, but all attempts failed. It would be helpful to have suggestions of input data format appear over the input text field for the new user. Thank you!

I have attached the video of the Owner role testing. This video includes user searching in the search bar, editing user data, interacting with the map, using user list functions, and displaying the count of countries.

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/69882989/5c157a60-7dde-4d5a-a014-094df8b71133

**Comment by palakgosalia** [comment]
Hey, I reviewed the code and tried testing by following the steps. I was able to modify users from the map as well as form the user list. I tried to add the user but I always got the Timezone key not found error even after adding the timezone variables as mentioned in the description. I used the Geocoding API key which is available for free. Please let me know if there is anything else that I need to do to make it run. Thank you


https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/104417272/16f0b916-8c68-4a40-b991-c7af9d668a9e

**Comment by xaanders** [comment]
> hi @xaanders the map renders perfect on my local environment and the number of countries show correctly this time. when in admin and volunteer login, the edit of team number are prohibited. Screenshot 2023-11-17 at 3 29 56 PM Screenshot 2023-11-17 at 3 31 41 PM when I try to create team on the local machine, it seems the new team is never saved. could you help me check how to create a team to test functions for onwers? Thank you!

There is no need to create a team to test functionality

**Comment by xaanders** [comment]
@ptrpengdev @palakgosalia 
Timezone key doesn't exist means that: 
a) it wasn't fetched on the dashboard page, currently it is the only place which make request for it (I will add fetching to this page too) 
b) it doesn't exist in the env file of backend

you can check it in the dev tools with redux dev tools if it exists in your global state

**Review by KavyaAlla — DISMISSED** [review]
Hi @xaanders I have tested this PR with PR 599 locally .
Review:
I successfully managed to add, edit, and delete users without encountering any errors. However, I noticed that sometimes, after navigating to other buttons and returning, the search and add user buttons are missing. Despite attempting to refresh the page, the issue persists(Video 2)


https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/61841929/2a5210ba-a5d2-4e9e-acf4-32ab5ce429b9


https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/61841929/69521cbf-8562-4043-9d9e-59d3e649775d

**Review by Shuhua-L — DISMISSED** [review]
Hey @xaanders ,

Looking good overall! 🙌 I’ve tested the following, and everything went smoothly:

- A user can be added through the “Add person” button.
- Existing users can be edited/deleted through the “Users list” or the searched results.
- Existing users can be found/edited/deleted through the map.
- The search bar works properly while inputting either names or locations.
- Updating an existing HGN user’s profile will update the map properly.
- Manager role users have no access to add/edit/remove/search users.

Not a big deal, but a potential minor improvement: The horizontal scroll bar within the search bar dropdown is not automatically showing up, making it a little bit hard to access the “Edit” buttons on the right-hand side.


https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/83729360/a12f3eee-fcd3-4d36-91ff-62856a24eaf2

File: src/components/TeamLocations/TeamLocations.jsx [hunk 1]
```diff
@@ -0,0 +1,317 @@
...
+
+function TeamLocations() {
+  const [userProfiles, setUserProfiles] = useState([]);
+  const [manuallyAddedProfiles, setManuallyAddedProfiles] = useState([]);
+  const [addNewIsOpen, setAddNewIsOpen] = useState(false);
+  const [listIsOpen, setListIsOpen] = useState(false);
+  const [editIsOpen, setEditIsOpen] = useState(false);
+  const [editingUser, setEditingUser] = useState(null);
+  const [searchText, setSearchText] = useState('');
+  const role = useSelector(state => state.userProfile.role);
```
Comment (robertoooc) [inline src/components/TeamLocations/TeamLocations.jsx:26]: Change this to: 
`` const role = useSelector(state => state.auth.user.role);``

**Review by robertoooc — DISMISSED** [review]
Hi @xaanders, I thoroughly tested this pr. All the functionality works well as intended the users' coordinates are reflected well on the map and it is editable as well. 

However, I realized a small flaw occurs when you're logged in as an Owner and then edit someone else's account in the slightest way. Then returning to the map the search bar and tools are gone although you are signed in as an owner. Looks like the redux state is storing the userprofile's role that you edited instead of the logged in user. 

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/117053202/727ef8a3-2745-478a-824a-b093d8c1714b

File: src/components/TeamLocations/TeamLocations.jsx (hunk 1 above)
Comment (xaanders) [reply src/components/TeamLocations/TeamLocations.jsx]: It is a good one! 
Thank you!

**Review by robertoooc — APPROVED** [review]
Great Job, everything works well now.

**Comment by beblicarl** [comment]
All functionality works. However, I think it would be a cool feature to search a user and automatically zoom to their saved location on the map when clicked on

[screencast-bpconcjcammlapcogcnnelfmaeghhagj-2023.11.24-14_04_03.webm](https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/68056104/f3d35cea-d0b1-4ae8-8cb5-d59cc6bd4af2)

**Review by beblicarl — APPROVED** [review]

**Review by navneeeth — APPROVED** [review]
Hello, @xaanders and @leonzh2k! Great implementation of this cool feature! The functionalities work as expected:

#### 1) Testing adding, editing, and removing the user from the map (marker/list) and from the existing user profile:

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/44264834/457b8e29-40d1-486e-9c7c-cc943de59ee9

#### 2) Creating a new user and testing functionality by punching 10+ hours and without punching hours:


https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/44264834/6fa5361d-5ad5-49a5-9a26-1e95a0caabe9

#### 3) Testing the functionality for Admin and Volunteer roles:


https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/44264834/15973506-6b74-4ce6-9872-84424544e4fd

#### 4) Search Functionality:
The search functionality works excellent in most aspects. However, I noticed that only the first name is searchable. In the following test, observe how I do not receive search results when I search for 'Oleksandr Volunteer' although there is an user on the map with that name. Perhaps in the next update, the functionality can have a search query that retrieves the full name in its results.


https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/44264834/34c115d1-8175-4b0e-9f52-9688a23b8bb9

Thank you!

**Pull request merged**